docs/adr/0035-refactor-payments-frontend.md

```diff
@@ -0,0 +1,101 @@
+# Refactor Subscription Platform frontend
+
+- Status: proposed
+- Deciders: Ben Bangert, Lisa Chan, Barry Chen
+- Date: 2023-03-07
+
+## Context and Problem Statement
+
+The Subscription Platform frontend has become overly complex, making it inflexible and difficult to work with. This has resulted in longer development times on new features and in some cases left us unable to easily accommodate requests from relying parties.
+
+Additionally, as the app has grown, so have the number of network calls causing a reduction in app performance. The aim is to refactor the app to have one network call for each action.
+
+Finally, the frontend includes some business and application logic, which has made components and parts of the app unnecessarily complex. As a result the frontend is less flexible, and harder to change and maintain. To solve this, the Subscription Platform team wants to move all business and application logic to the backend.
+
+To address the problems briefly discussed above, and to meet current and future relying party and feature requests, refactoring of the existing Subscription Platform is necessary.
+
+## Decision Drivers
+
+- Faster development - This applies to both the refactoring necessary for SubPlat 3.0, as well as future relying party and feature requests
+- Easier routing - The SubPlat teams wants to perform more advanced routing to allow features such as, Relying Party specific checkout pages, A/B testing, etc,
+- GraphQL support - Improve app performance and responsivness by reducing the number of network requests
+- Reuse existing components - We have received generally favourable reviews of our existing frontend look and feel, and want to keep the appearance the same
+
+## Considered Options
+
+- A. React Framework - Next.js
+- B. React Framework - Gatsby.js
+- C. React, Typescript, without Redux
+- D. React, Typescript, with Redux
+- E. React Framework - Remix
+
+## Decision Outcome
+
+Chosen option: "A. React Framework - Next.js", because a significant amount of refactoring is necessary to accomplish the Subscription Platforms goals, now would be a good time to adopt a React framework and start from a mostly clean slate, since it partially or completely meets all of the decision drivers, especially the easier routing and faster development drivers.
+
+Regarding why Next.js is favored over Gatsby.js, before completing the ADR, the subscription platform team went through a prototyping project and found the two frameworks to be remarkably similar in capabilities. However due to Next.js's popularity, use by other teams in Mozilla and a team vote, the decision was made to pick Next.js.
+
+## Pros and Cons of the Options
+
+### A. React Framework - Next.js
+
+Use a React framework, Next.js.
+
+- Good, because it provides many benefits out of the box, without compromising flexibility, that would require additional libraries otherwise. (Routing, rendering options, image optimization, etc.)
```

Comment:
We probably want to mention server-side rendering here right?

It could be worth referencing [this ADR](https://github.com/mozilla/fxa/blob/main/docs/adr/0010-transition-fxa-from-backbone-to-react.md#option-c---move-to-isomorphic-react) for some pros or even that explanation up top of what SSR is. Also probably worth linking to it and explaining what we've learned since then (though granted that ADR was for fxa settings and not subplat).

Reply:
At the moment SSR is indirectly mentioned by `rendering options`.

SSR wasn't really a decision driver for this ADR. It's more that the framework offers a lot out of the box, which allows for faster development.

Reply:
Without SSR, wouldn't moving logic to the server (mentioned here) be more difficult since it'd require the client to spin up and then send requests out vs sending data back with the first render? The flexibility at least is nice.

Reply:
Oh I think I see where the mix up might be happening. When the ADR references `...move all business and application logic to the backend...`, it doesn't necessarily mean the Next.js server, but also the payments APIs in `fxa-auth`.

A good example of this is the logic in the `'/oauth/mozilla-subscriptions/customer/plan-eligibility/{planId}'` API in `fxa-auth`, which used to be done in React.

The reason why this is mentioned in the ADR is to try illustrate that the frontend refactor will change/simplify quite a bit of the frontend, making it an ideal time to adopt a framework. It wasn't necessarily added to indicate that we want a SSR solution. (All that said, at the moment it does look like we'll be SSR first for our pages, and might have some exceptions)

Reply:
Ah, that makes more sense. Maybe that's good to clarify then - when I think of NextJS, the two things that immediately come to mind are "React framework" and "SSR" (thirdly SSG). I suppose IMO if I was looking at this ADR in the future I might want to see what we thought about SSR at the time and might make assumptions reading that some logic would be moved to the back-end that it would mean the Next back-end, as I'm guessing there's at least a couple scenarios where sending down data on `getInitialProps` or `getServerSideProps` would be beneficial. I know it's not the primary reason for the ADR so I'll leave it up to you on what's worth noting vs not noting. Maybe I'm just projecting since if/when FxA uses Next we'll definitely have a note about SSR in an ADR since there's been a couple places I've seen during our Backbone to React refactoring work where it would help us out.